packages/backend/lambda-utils/src/build-api-response.ts

```diff
@@ -12,10 +15,39 @@ import { CustomAPIGatewayProxyEvent } from "./gateway-proxy-handler";
    * @param event An AWS API Gateway proxy event object
    * @param event An AWS API Gateway proxy event object
    * @param data A JSON API response object
+   * @param allowedOrigins A collection of origins to allow
    */
 export const buildApiResponse = <D extends unknown>(
-  event: CustomAPIGatewayProxyEvent, _context: Context, data: D | APIError<unknown> | Error
+  event: CustomAPIGatewayProxyEvent, _context: Context, data: D | APIError<unknown> | Error, allowedOrigins?: string[]
 ): APIGatewayProxyResult => {
+  if (allowedOrigins) {
+    const currentOrigin = event.headers ? event.headers.origin : null;
+    let isOriginAllowed = false;
+
+    // Check if origin is in allowed origins
+    allowedOrigins.map(allowedOrigin => {
+      if (!isOriginAllowed && currentOrigin && currentOrigin.match(allowedOrigin)) {
+        isOriginAllowed = true;
+      }
+    });
+
+    // Return an error if origin is not allowed or if there's no origin
+    if (!isOriginAllowed || !currentOrigin) {
+      const error = new APIError("Forbidden - origin not allowed", ErrorType.ForbiddenError);
+      const { statusCode, ...formattedError } = APIError.formatApiError(error);
+
+      return {
+        statusCode,
+        headers: {
+          "Content-Type": "application/json",
+          "Access-Control-Allow-Origin": allowedOrigins[ 0 ],
```

Comment:
actually just thinking you should return the currentOrigin here as it has passed validation. We could get issues if the header doesn't match the request otherwise

Reply:
I think that should remain `allowedOrigins[0]` since it's what gets returned if `!isOriginAllowed || !currentOrigin`.

Reply:
Did you mean to return `currentOrigin` for this bit (after the check)
```
const headers = {
    "Content-Type": "application/json",
    "Access-Control-Allow-Origin": event.headers && event.headers.origin ? event.headers.origin : "*",
    "Access-Control-Allow-Credentials": true
  };
```